### PR TITLE
Windows specific patches

### PR DESCRIPTION
**Pull request by ashfinal**
So I finally get time trying to address issue #133. These patches include:

1. ~~Use subprocess.Popen instead of os.fork. As os.fork is dropped in Python 3.8, and doesn't work on Windows, I believe using subprocess.Popen is a better choice.~~ Use multiprocessing now, see #136 
2. Default tcp address: `127.0.0.1:6789` on Windows. Before build 17063 Windows doesn't support unix socket, so we use a tcp address. On *nix system, still use `/tmp/nvimsocket`.
3.  Fix `--serverlist` on Windows. `--serverlist` arg didn't work on Windows, this fixes it.
4. Also produce `gnvr` binary for GUI use. If you want to right-click something in file explorer and open it, just use `gnvr`. The difference from `nvr` is it doesn't bring up the ugly command line window if you are working on GUI desktop.
5. Use double quotes in doc. It seems on Windows args parsing is more strict/limited, outer single quotes usually are considered as errors. We use double quotes to avoid confusion.
6. Make noise only if user makes mistakes. i.e. `nvr` now should work out of the box. If no neovim instance is found, it will automatically help launch one(and do the rest work in queue, remote-expr, remote-send... etc). It also launches another instance if user explicitly designate an address with `--servername` to attach(but without success), unless `--nostart` is also given. This makes `nvr` flexible enough to run multiple neovim instances, and user friendly.

These patches also obsolete my `invr.au3` script.

Let me know if I break something :)

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -17,7 +17,8 @@
     python_requires  = '>=3.5',
     install_requires = ['pynvim', 'psutil', 'setuptools'],
     entry_points     = {
-        'console_scripts': ['nvr = nvr.nvr:main']
+        'console_scripts': ['nvr = nvr.nvr:main'],
+        'gui_scripts': ['gnvr = nvr.nvr:main']
```
Comment (mhinz) [inline setup.py]: What is this for? Why is it needed?

File: nvr/nvr.py [hunk 2]
```diff
@@ -65,8 +66,15 @@ def start_new_process(self, silent):
         args = os.environ.get('NVR_CMD')
         args = args.split(' ') if args else ['nvim']
 
-        pid = os.fork()
-        if pid == 0:
+        os.environ['NVIM_LISTEN_ADDRESS'] = self.address
+
+        try:
+            pid = subprocess.Popen(args).pid
```
Comment (mhinz) [inline nvr/nvr.py]: Unfortunately this leaves me with a hanging shell. Does this really work on Windows?

Let me read up on `subprocess` and play around with it. Maybe it just needs another argument.

**Comment by mhinz** [comment]
Please drop the double-quote commit, since I don't see the technical merit. Let's concentrate on the process handling for now. :)

**Comment by mhinz** [comment]
I wonder if `subprocess` is even applicable in this case. On Unix, it does use `os.fork()` and `os.execvp()` in the child, but that makes `nvim` spawn in the _background_ without any way to get it into the foreground. But that's what we want. If `nvr` finishes, we should be in a nvim process.

Maybe the correct approach would be to replace the current process via `os.execvpe()` (like master does), but somehow use `multiprocessing` (?) before that to keep nvr running in the background (instead of the other way around).

**Comment by mhinz** [comment]
The issue I get on Unix with this PR is the same as if I'd using master but switching the bodies of `if/else` here: https://github.com/mhinz/neovim-remote/blob/master/nvr/nvr.py#L69-L86

File: setup.py (hunk 1 above)
Comment (ashfinal) [reply setup.py]: `gui_scritpts` will create an binary `gnvr.exe`, the difference from `nvr.exe` is it wouldn't bring up the terminal. I guess on Windows there is a clear line between "console program" and "gui program".  While on *nix, `gnvr` and `nvr` are just the same script. The user case is mentioned as above.

File: nvr/nvr.py (hunk 2 above)
Comment (ashfinal) [reply nvr/nvr.py]: As I've set `NVR_CMD` to `nvim-qt`, it worked great that I forgot to check about the situation under terminal. :) Anyway, I did a little investigation, it seems under terminal when subprocess.Popen run nvim background, it'll scrape your terminal or cause it hanging. 

I think this is because nvim is designed as an interactive program. If we pass the argument `--headless` to subprocess.Popen, the problem is solved! But that's not what we want. We need nvim to stay acting interactively when we start an new process for users. 

I did more investigation, and came to the conclusion that maybe the simplest solution is we just  launch nvim and let it wait. I will create a commit based on this.

**Comment by ashfinal** [comment]
@mhinz I've made some improvement, and did some test under alpinelinux(clearly not enough). Can you pull the latest commits and see if things get better?

File: setup.py (hunk 1 above)
Comment (mhinz) [reply setup.py]: What happens if you use `gnvr.exe` with `NVR_CMD=nvim` (the default)?

I think `gui_scripts` is only needed for Windows, but the question is: Do we even need `console_scripts`?

File: nvr/nvr.py (hunk 2 above)
Comment (mhinz) [reply nvr/nvr.py]: But it's likely that nvr starts a long-running nvim process and it feels wrong to have nvr wait in that case. If the work is done, nvr should exit and nvim should continue running.

**Comment by mhinz** [comment]
Could you try master again?

The latest commit replaces `os.fork()` by multiprocessing and does the opposite of what subprocess would be doing: It replaces the current process with the new nvim process and before that starts a new child process that makes nvr attach to the new process, does its thing, and then exits.

_Theoretically_, this should work for Win as well. If it does, I'll cherry-pick your other commits from this PR onto master.

File: setup.py (hunk 1 above)
Comment (ashfinal) [reply setup.py]: Set `NVR_CMD=nvim`, and type `gnvr.exe file` in your terminal, it will still work(partly), except that it will NEVER echo any messages, ever it encounts huge errors.

We'll still need `console_scripts`, they are respectively for different user cases.

File: nvr/nvr.py (hunk 2 above)
Comment (ashfinal) [reply nvr/nvr.py]: Yeah, I once thought if there was any way we could just let python exit and nvim continue running.  If possible, that would be great

**Comment by ashfinal** [comment]
Emm, the lastest commit says:  parameter error.
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Program Files\Python38-32\lib\multiprocessing\spawn.py", line 102, in spawn_main
    source_process = _winapi.OpenProcess(
OSError: [WinError 87] 参数错误。
```
Maybe you pass wrong paremeters? I will investigate the code.

**Comment by mhinz** [comment]
Maybe it's a Python bug? https://bugs.python.org/issue37549

**Comment by ashfinal** [comment]
Looks ok to me on Windows 10.
```python
>>> import os
>>> os.dup(1)
3
>>> 
````

**Comment by mhinz** [comment]
Hmm, there are so many issues when you search for `OSError: [WinError 87]`... why can't stuff just work?

**Comment by ashfinal** [comment]
I think I kinda figure it out. `os.execvpe` will replace the current process immediately, open file objects and descriptors are not flushed. From what i see, unless we are 100% sure the job is done, we are never supposed to replace the current process. If we have to use `os.execvpe`, it'd better be our last step or in another python interpreter process, not in main loop.

I'll create a PR,  please pull it and see if it works.

**Comment by ashfinal** [comment]
@mhinz I revert subprocess.Popen related commits, and make minor fixes about `show_message` function. 

Please merge #136 first, then this PR will be ready to be merged. :)

**Comment by jpboyerrr** [comment]
Thanks to this pull request (and #136), I now have access to a core functionality that previously stalled my full workflow migration from gVim to Neovim on Windows 10. However, I did have to manually update nvr.py to ensure the nvim-qt process is called upon instead of the default terminal nvim process. As a layman, I do not know how to best handle all scenarios graciously, but I do think the end-user shouldn't have to recompile code based off whether her/his workflow is nvim-centric or nvim-qt-centric. I would have expected nvr and gnvr to be different in that respect; both call the same process from what I currently understand.

See the following line where I updated 'nvim' to 'nvim-qt' to support my specific workflow:
https://github.com/ashfinal/neovim-remote/blob/100a1f854ee96cca82fa3d19b1b06fd94acfd0e2/nvr/nvr.py#L65

Thanks again!

**Comment by ashfinal** [comment]
Glad to hear that! :) That's nice.

Did you also merge #136 or just #135 alone? These two PRs should make neovim-remote work on Windows and Unix platforms.

Tips: you don't have to recompile code to be able to run nvim-qt. You just need to set environment variable 'NVR_CMD' to 'nvim-qt'. And, nvr and gnvr are the same thing for the most part, except gnvr doesn't depend on Command Prompt, while nvr does. When using  context menu's "Edit with Neovim-qt" in Explorer, you certainly don't want to see the big black Command Prompt.

**Comment by jpboyerrr** [comment]
Thanks for the tip, I must have read the README a bit too literally. Might I suggest rewording the following passage to confirm NVR_CMD can be used on any platform ?

`If the targeted address does not exist, nvr starts a new process by running "nvim". You can change the command by setting $NVR_CMD. (This requires forking, so it won't work on Windows.)`

FYI, I did merge #136, then #135 as requested above. Setting the proper registry for file right click support worked flawlessly when using either nvr or gnvr as the called process. Therefore I didn't have the black window you mentioned even using nvr, just the actual nvim terminal.

**Comment by ashfinal** [comment]
Sure, we should definitely change that paragraph since neovim-remote will work on most platforms, of course, after @mhinz has merged these two PRs. On Windows, the environment variable is `%NVR_CMD%`.

d

> Therefore I didn't have the black window you mentioned even using nvr, just the actual nvim terminal.

That's strange. When using nvr, you should see Command Prompt window flashing by very quickly. Maybe it's related to specific version? My Windows version is 10.0.14393, and Python version is 3.8.1. Anyway, it's a tiny flaw, should not matter much.

**Comment by luiq54** [comment]
@ashfinal Thanks! I am using your branch directly and it works flawlessly!
